File: Corrections/Control_Segment/getCtrl_corr.m
function    [X, tcorr,tgd]   =   getCtrl_corr(eph, svn, TOW, pr,str)
% getSatPos_tcorr:  Get satellite coordinates and clock correction.  
%
% Inputs:
%           - eph:  Ephemeris data obtained from the Navigation RINEX file
%                   cointaining the data needed for the computation of the
%                   satellite coordinates and clock corrections.
%           - svn:  Integer with the sat. ID of the satellite to be analyzed
%           - TOW:  TOW of the epoch to analyze.
%           - pr:   Pseudorange measure at given TOW
% Outputs:
%           - X:        3x1 matrix with the corrected cartesian coordinates 
%                       of the satellite with Id given by svn at the given 
%                       TOW.
%           - tcorr:    Clock correction in [sec.] for the satellite with  
%                       Id given by svn at the given TOW.
%

    %
    c           =   299792458;          %   Speed of light (m/s)
    %-  Get satellite coordinates and clock corrections
    tx_RAW      =   TOW - pr/c;         %   1st guess of Tx time
    %       
    %--     Identify ephemerides columns in eph
    col         =   find_eph(eph, svn, tx_RAW);
    %
    %--     Get clock corrections
    tcorr       =   sat_clock_error_correction(tx_RAW, eph(:, col));
    tgd         =   eph(22, col);
%     tcorr       =   tcorr - tgd;        %   Correct the total Group Delay (TGD)
    tx_GPS      =   tx_RAW-tcorr;       %   Correct the Tx time
    %--     Compute again the clock bias
    tcorr       =   sat_clock_error_correction(tx_GPS, eph(:, col));
%     tcorr       =   tcorr - tgd;    
    %
    %-  Get satellite coordinates (corrected) and velocity
    [X, vel]     =   satpos(tx_GPS, eph(:, col),str);    
    %
    %--     Get the satellite relativistic clock correction
    trel        =   -2 * ( dot(X, vel) / (c^2) ); % IS-GPS-200E, p. 86

    %
    %--     Account for the relativistic effect on the satellite clock bias
    %       and the time of transmission
    tcorr       =   tcorr + trel;   
    tx_GPS      =   tx_RAW - tcorr;  % Corrected GPS time
    %
    %--     Recompute the satellite coordinates with the corrected time and
    %       some additional correction (i.e. Sagnac effect / rotation
    %       correction
    traveltime  =   TOW - tx_GPS;
    X           =   e_r_corr(traveltime, X);

end

File: pvt-computation/src/main/java/com/abecerra/pvt_computation/domain/computation/corrections/GetCtrlCorr.kt
package com.abecerra.pvt_computation.domain.computation.corrections

import com.abecerra.pvt_computation.data.EcefLocation
import com.abecerra.pvt_computation.data.PvtConstants
import com.abecerra.pvt_computation.data.input.SatelliteMeasurements
import com.abecerra.pvt_computation.domain.computation.utils.satPos
import com.abecerra.pvt_computation.data.PvtConstants.C
import com.abecerra.pvt_computation.domain.computation.utils.checkTime
import com.abecerra.pvt_computation.domain.computation.utils.earthRotCorr
import org.ejml.data.DMatrixRMaj
import org.ejml.dense.row.CommonOps_DDRM
import kotlin.math.pow

fun getCtrlCorr(
    satellite: SatelliteMeasurements,
    tow: Double,
    pR: Double
): CtrlCorr {

    // Compute transmission time
    val txRaw = tow - pR / C

    // Get clock corrections
    var tCorr = satClockErrorCorrection(txRaw, satellite)

    tCorr -= getTgds(satellite)

    var txGPS = txRaw - tCorr

    // Compute again the clock bias
    tCorr = satClockErrorCorrection(txGPS, satellite)
    tCorr -= satellite.satelliteEphemeris.tgdS

    // Get the satellite coordinates (corrected) and velocity
    val satPos = satPos(txGPS, satellite)
    var x = satPos.x
    val vel = satPos.vel

    val xVec = DMatrixRMaj.wrap(3, 1, x)
    val velVec = DMatrixRMaj.wrap(3, 1, vel)

    // Get the satellite relativistic clock correction
    val tRel: Double = -2 * (CommonOps_DDRM.dot(xVec, velVec) / (C * C))

    // Account for the relativistic effect on the satellite clock bias and the time of transmission
    tCorr += tRel
    txGPS = txRaw - tCorr

    // Recompute the satellite coordinates with the corrected time and some additional
    // correction (i.e. Sagnac effect / rotation correction)
    val travelTime = tow - txGPS
    x = earthRotCorr(travelTime, x)

    return CtrlCorr(
        EcefLocation(x[0], x[1], x[2]),
        tCorr
    )
}

private fun getTgds(satellite: SatelliteMeasurements): Double {
    return if (satellite.constellation == PvtConstants.GALILEO
        && satellite.carrierFreq == PvtConstants.L5_FREQ
    ) satellite.satelliteEphemeris.tgdS * (PvtConstants.L1_FREQ / PvtConstants.L5_FREQ).pow(2)
    else satellite.satelliteEphemeris.tgdS
}

fun satClockErrorCorrection(time: Double, satellite: SatelliteMeasurements): Double {
    var corr: Double
    with(satellite.satelliteEphemeris) {
        var dt = 0.0
        this.keplerModel?.let {
            dt = checkTime(time - it.toeS)
        }
        corr = (af2 * dt + af1) * dt + af0
    }
    return corr
}

data class CtrlCorr(
    var ecefLocation: EcefLocation,
    var tCorr: Double = 0.0
)
